proposed/resource-limit/resource-limit.md

```diff
@@ -84,6 +84,35 @@ TDB
 
 The `Resource` struct is designed to handle the releasing of the resource. This way, the user won't be able to release more resources than was obtained, unlike Semaphores. We've also identified scenarios where additional information need to be returned by the limiter, such as reason phrases, response codes, percentage of rate saturation, retry after etc. For these cases, the object `Resource.State` can be used to store the additional metadata.
 
+The motivation behind the `TryAcquire` and `AcquireAsync` distinction is that there are use cases where a fast synchronous check is required and there are other cases where waiting until a resource is available is better suited. There are parallels for this in System.Threading.Channels with ChannelWriter.TryWrite vs ChannelWriter.WriteAsync. We envision `AcquireAsync` to also allow for waiting for pending acquisition requests (FIFO or LIFO) when resources are exhausted. In these cases, the behaviour is controlled with two limits:
+
+1. Resource limit: this represents how many resource are available to be acquired immediately
+2. Queue/Stack limit: this represents how many requests can be queued
+
+This gives 3 scenarios :
+
+1. Count + RequestedCount <= Resource limit: In this case `TryAcquire` will return `true` and `AcquireAsync` will return immediately
+2. Resource limit < Count + RequestedCount <= Resource limit + Queue/Stack limit: In this case `TryAcquire` will return `false` and `AcquireAsync` will wait in a queue.
+3. Count + RequestedCount > Resource limit + Queue/Stack limit: In this case `TryAcquire` will return `false` and `AcquireAsync` will throw an exception immediately.
+
+In case we want to reduce the allocation of an `Exception` type in scenario 3, we can return a `bool` either as a tuple for `AcquireAsync`, i.e. `ValueTask<(bool, Resource)> AcquireAsync(long requestedCount, CancellationToken cancellationToken = default)`, or add a `bool` to the `Resource` struct.
```

Comment:
Yes please. I think this could also look like a deconstruct method on Resource:

```C#
public struct Resource : IDisposable
{
    ...
    
    public Resource(object? state, Action<Resource>? onDispose, bool success)
    {
        State = state;
        _onDispose = onDispose;
        _success = success;
    }
    public void Deconstruct(out Resource resource, out bool success)
    {
          resource = this;
          success = _success;
    }

    public static Resource Success = new Resource(null, null, success: true);
    public static Resource Fail = new Resource(null, null, success: false);
}
```

Reply:
This has been adopted in the proposal